geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/BinnedTime.scala

```diff
@@ -0,0 +1,170 @@
+  *     bin      => year
+  *     offset   => minutes
+  *     max date => 01/01/34737
+  *
+  * @param bin number of time periods from the java epoch
+  * @param offset precise offset into the specific time period
+  */
+case class BinnedTime(bin: Short, offset: Long)
+
+object BinnedTime {
+
+  val Epoch = new DateTime(0, DateTimeZone.UTC)
+
+  type TimeToBinnedTime = (Long) => BinnedTime
+  type DateToBinnedTime = (DateTime) => BinnedTime
+  type BinnedTimeToDate = (BinnedTime) => DateTime
+
+  /**
+    * Gets period index (e.g. weeks since the epoch) and offset into that interval (e.g. seconds in week)
+    *
+    * @param period interval type
+    * @return
+    */
+  def timeToBinnedTime(period: TimePeriod): TimeToBinnedTime = {
+    period match {
+      case TimePeriod.Day   => toDayAndMillis
+      case TimePeriod.Week  => toWeekAndSeconds
+      case TimePeriod.Month => toMonthAndSeconds
+      case TimePeriod.Year  => toYearAndMinutes
+    }
+  }
+
+  /**
+    * Gets period index (e.g. weeks since the epoch) and offset into that interval (e.g. seconds in week)
+    *
+    * @param period interval type
+    * @return
+    */
+  def dateToBinnedTime(period: TimePeriod): DateToBinnedTime = {
+    period match {
+      case TimePeriod.Day   => toDayAndMillis
+      case TimePeriod.Week  => toWeekAndSeconds
+      case TimePeriod.Month => toMonthAndSeconds
+      case TimePeriod.Year  => toYearAndMinutes
+    }
+  }
+
+  /**
+    * Gets a date back from a binned time
+    *
+    * @param period interval type
+    * @return
+    */
+  def binnedTimeToDate(period: TimePeriod): BinnedTimeToDate = {
+    period match {
+      case TimePeriod.Day   => fromDayAndMillis
+      case TimePeriod.Week  => fromWeekAndSeconds
+      case TimePeriod.Month => fromMonthAndSeconds
+      case TimePeriod.Year  => fromYearAndMinutes
+    }
+  }
+
+  /**
+    * Gets the max offset value for a given time period
+    *
+    * @param period interval type
+    * @return
+    */
+  def maxOffset(period: TimePeriod): Long = {
+    period match {
+      case TimePeriod.Day   => Days.ONE.toStandardDuration.getMillis
+      case TimePeriod.Week  => Weeks.ONE.toStandardDuration.getMillis / 1000L
+      case TimePeriod.Month => (Days.ONE.toStandardDuration.getMillis / 1000L) * 31L
+      case TimePeriod.Year  => (Weeks.ONE.toStandardDuration.getMillis / 60000L) * 52L
+    }
+  }
+
+  private def toDayAndMillis(time: Long): BinnedTime =
+    toDayAndMillis(new DateTime(time, DateTimeZone.UTC))
+
+  private def toDayAndMillis(date: DateTime): BinnedTime = {
+    val days = Days.daysBetween(Epoch, date)
+    val millisInDay = date.getMillis - Epoch.plusDays(days.getDays).getMillis
+    BinnedTime(days.getDays.toShort, millisInDay)
+  }
+
+  private def fromDayAndMillis(date: BinnedTime): DateTime = Epoch.plusDays(date.bin).plus(date.offset)
+
+  private def toWeekAndSeconds(time: Long): BinnedTime =
+    toWeekAndSeconds(new DateTime(time, DateTimeZone.UTC))
```

Comment:
Joda object creation isn't free; any chance that it'd make sense to do this arithmetic manually?